crmprtd/align.py

```diff
@@ -8,12 +8,237 @@
+
+        if obs_tuple.lat and obs_tuple.lon:
+            matching_stns = []
+            for hid in q.first():
+                matching_stns.append(hid)
+            matching_stns = set(matching_stns)
+            close_stns = matching_stns.intersection(
+                closest_stns_within_threshold(sesh, obs_tuple.lon,
+                                              obs_tuple.lat, 800))
+            log.debug('Station set intersection',
+                      extra={'station_set': close_stns})
+
+            if len(close_stns) == 0:
+                log.debug('Station set is empty, adding new station')
+                hid = create_station_and_history_entry(sesh, obs_tuple)
+
+            elif len(close_stns) == 1:
+                log.debug('Single station in set')
+                hid = close_stns.pop()
+
+            elif len(close_stns) > 1:
+                log.debug('Multiple stations in set')
+                closest_pos = 999   # max distance should be 800m at this point
+                for id in close_stns:
+                    try:
+                        lat, lon = sesh.query(History.lat, History.lon).filter(
+                            History.id == id).first()
+                    except Exception as e:
+                        log.warning('An error occured while accesing the db',
+                                    extra={'exception': e})
+
+                    dist = haversine_formula(obs_tuple.lat,
+                                             obs_tuple.lon,
+                                             float(lat),
+                                             float(lon))
+
+                    # best case, we find exact match
+                    if dist == 0:
+                        log.debug('Exact match found')
+                        hid = id
+                        break
+
+                    # find closest
+                    else:
+                        if dist < closest_pos:
+                            hid = id
+                            closest_pos = dist
+
+            if not hid:
+                log.debug('Observation contains lat/lon but does not match '
+                          'any entries in db')
+                hid = create_station_and_history_entry(sesh, obs_tuple)
+        else:
+            log.debug('No lat/lon, checking for \"active\" stations')
+            for id in q.all():
+                q = sesh.query(History.id).filter(
+                    and_(History.id == id,
+                         History.sdate is not None,
+                         History.edate is None))
+
+                if q.count() > 0:
+                    log.debug('Matched hid using sdate/edate')
+                    hid, = q.first()
+                    break
+
+        if not hid:
+            log.warning('Unable to match station',
+                        extra={'station_id': obs_tuple.station_id,
+                               'network_name': obs_tuple.network_name})
+            return
+
+    # thing
```

Comment:
So I know that *I* called this a "thing" in our introductory discussion about this, but it was simply because I couldn't think of a better way to describe it. We should be more specific here. "Quantity"? "Variable"? "Measurement"?